### PR TITLE
Create S2S Auth Client

### PR DESCRIPTION
**Pull request by dharmendrak**

### CODE REVIEW

File: src/main/resources/META-INF/spring.factories [hunk 1]
```diff
@@ -0,0 +1 @@
+org.springframework.boot.autoconfigure.EnableAutoConfiguration=uk.gov.hmcts.authorisation.config.ServiceTokenGeneratorConfiguration,uk.gov.hmcts.authorisation.ServiceAuthorisationApi
```
Comment (rkondratowicz) [inline src/main/resources/META-INF/spring.factories]: what is this file for?

File: build.gradle [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    id 'io.spring.dependency-management' version '1.0.3.RELEASE'
+    id 'org.springframework.boot' version '1.5.8.RELEASE'
+}
+
+apply plugin: 'checkstyle'
+apply plugin: 'java-library'
+apply plugin: 'org.springframework.boot'
+apply plugin: 'maven'
+apply plugin: 'maven-publish'
+apply plugin: 'com.jfrog.bintray'
```
Comment (rkondratowicz) [inline build.gradle]: can those go to the block above?

File: build.gradle [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        }
+    }
+}
+
+ext {
+    springCloudVersion = 'Dalston.SR4'
+}
+
+dependencies {
+    compile('org.springframework.cloud:spring-cloud-starter-feign')
```
Comment (rkondratowicz) [inline build.gradle]: Can we use vanilla feign instead?

File: build.gradle [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    compile('com.netflix.feign:feign-jackson:8.18.0')
+    compile group: 'com.google.guava', name: 'guava', version: '23.0'
+    compile group:'com.warrenstrange', name: 'googleauth', version: '1.1.2'
+    testCompile group: 'org.springframework.boot', name: 'spring-boot-starter-test', version: springBootVersion
+    testCompile group: 'org.assertj', name: 'assertj-core', version: '3.8.0'
+    testCompile group: 'com.marvinformatics.feign', name: 'feign-mock', version: '0.6'
+    testCompile group: 'io.github.openfeign', name: 'feign-gson', version: '9.4.0'
```
Comment (rkondratowicz) [inline build.gradle]: jackson at runtime and gson in tests?

File: src/main/java/uk/gov/hmcts/authorisation/config/ServiceTokenGeneratorConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    @Bean
+    public AuthTokenGenerator cachedServiceAuthTokenGenerator(
+            @Qualifier("serviceAuthTokenGenerator") AuthTokenGenerator serviceAuthTokenGenerator,
+            @Value("${idam.s2s-auth.tokenTimeToLiveInSeconds:14400}"
+            ) int ttl) {
+        return new CachedServiceAuthTokenGenerator(serviceAuthTokenGenerator, ttl);
+    }
+}
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/authorisation/config/ServiceTokenGeneratorConfiguration.java]: what is this class for?

File: src/main/java/uk/gov/hmcts/authorisation/generators/ServiceAuthTokenGenerator.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+            final ServiceAuthorisationApi serviceAuthorisationApi
+    ) {
+        this.secret = secret;
+        this.microService = microService;
+        this.serviceAuthorisationApi = serviceAuthorisationApi;
+    }
+
+    @Override
+    public String generate() {
+        final GoogleAuthenticator googleAuthenticator = new GoogleAuthenticator();
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/authorisation/generators/ServiceAuthTokenGenerator.java]: can be initialized once in constructor or field declaration.

File: .gitignore [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+# Avoid ignoring Gradle wrapper jar file (.jar files are usually ignored)
+!gradle-wrapper.jar
+
+# Cache of project
+.gradletasknamecache
+
+# # Work around https://youtrack.jetbrains.com/issue/IDEA-116898
+# gradle/wrapper/gradle-wrapper.properties
+
+# End of https://www.gitignore.io/api/java,macos,gradle,eclipse,windows,intellij+all
```
Comment (rkondratowicz) [inline .gitignore]: would that be enough? https://github.com/hmcts/java-logging/blob/master/.gitignore (16 lines):)

File: src/main/java/uk/gov/hmcts/authorisation/generators/ServiceAuthTokenGenerator.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        final String oneTimePassword = valueOf(googleAuthenticator.getTotpPassword(secret)).toString();
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/authorisation/generators/ServiceAuthTokenGenerator.java]: this convert `int` -> `Integer` -> `String`

Use either `String.valueOf(i)` or `Integer.toString(i)` instead.

Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/authorisation/generators/ServiceAuthTokenGenerator.java]: also, does it not need to be padded with leading zeros?

File: src/main/java/uk/gov/hmcts/authorisation/config/ServiceTokenGeneratorConfiguration.java (hunk 5 above)
Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/authorisation/config/ServiceTokenGeneratorConfiguration.java]: this caches the token

File: src/main/java/uk/gov/hmcts/authorisation/generators/ServiceAuthTokenGenerator.java (hunk 8 above)
Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/authorisation/generators/ServiceAuthTokenGenerator.java]: reply it doesn't need, it works as it

File: build.gradle (hunk 4 above)
Comment (dharmendrak) [reply build.gradle]: removed, we add more testing later when I have enough confidence testing feign clients

File: src/main/java/uk/gov/hmcts/authorisation/config/ServiceTokenGeneratorConfiguration.java (hunk 5 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/authorisation/config/ServiceTokenGeneratorConfiguration.java]: I mean `ServiceTokenGeneratorConfiguration` class

File: build.gradle (hunk 3 above)
Comment (dharmendrak) [reply build.gradle]: for some reason, it is followed to use cloud one across all feign clients. with vanilla feign lot of testing strategy available

Comment (rkondratowicz) [reply build.gradle]: We use spring starter feign in spring boot applications. No need to use spring in a library.

File: build.gradle [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+compileTestJava {
+    options.compilerArgs << '-Xlint:deprecation'
+}
+
+repositories {
+    maven {
+        url "https://dl.bintray.com/hmcts/hmcts-maven"
+    }
+    jcenter()
+    mavenLocal()
```
Comment (doncem) [inline build.gradle]: only needed for local development - perhaps avoid committing?

File: build.gradle [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    }
+}
+
+task printVersion {
+    doLast {
+        print project.version
+    }
+}
+
+def springBootVersion = '1.5.7.RELEASE'
```
Comment (doncem) [inline build.gradle]: resolve version mismatch by adding `plugins.getPlugin('org.springframework.boot').class.package.implementationVersion`, see spring boot template

File: build.gradle [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+        licenses = ['MIT']
+        vcsUrl = 'https://github.com/hmcts/service-authorisation-access-client'
+        version {
+            name = project.version
+        }
+    }
+}
+
+ext {
+    springCloudVersion = 'Dalston.SR4'
```
Comment (doncem) [inline build.gradle]: what kind of version is that?

> googling... oh dear

File: gradle/wrapper/gradle-wrapper.properties [hunk 12]
```diff
@@ -0,0 +1,6 @@
+#Mon Dec 04 11:50:49 GMT 2017
+distributionBase=GRADLE_USER_HOME
+distributionPath=wrapper/dists
+zipStoreBase=GRADLE_USER_HOME
+zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-3.5-all.zip
```
Comment (doncem) [inline gradle/wrapper/gradle-wrapper.properties]: `4.3.1` is currently in use

File: src/main/java/uk/gov/hmcts/authorisation/config/ServiceTokenGeneratorConfiguration.java (hunk 5 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/authorisation/config/ServiceTokenGeneratorConfiguration.java]: if yes, could introduce `ConfigurationProperties` interface and so reduce the character clutter within bean definition

File: build.gradle [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+apply plugin: 'checkstyle'
+apply plugin: 'java-library'
+apply plugin: 'org.springframework.boot'
+apply plugin: 'maven'
+apply plugin: 'maven-publish'
+apply plugin: 'com.jfrog.bintray'
+
+group 'uk.gov.hmcts.reform'
```
Comment (doncem) [inline build.gradle]: mismatching with file structure: is it going under platform reform group or just generic hmcts authorisation package?

File: src/main/java/uk/gov/hmcts/authorisation/config/ServiceTokenGeneratorConfiguration.java (hunk 5 above)
Comment (dharmendrak) [reply src/main/java/uk/gov/hmcts/authorisation/config/ServiceTokenGeneratorConfiguration.java]: done

File: build.gradle (hunk 9 above)
Comment (dharmendrak) [reply build.gradle]: we can take it out once we are happy, however it helps in testing in actual use

File: build.gradle (hunk 3 above)
Comment (timja-kainos) [reply build.gradle]: we only use spring here though, easier imo to just stay in the spring eco system. 
but not convinced either way which is better.

File: build.gradle (hunk 9 above)
Comment (timja-kainos) [reply build.gradle]: why would only needed for local development be a reason not to commit it?
it makes it much easier for local, I would recommend keeping it

File: build.gradle (hunk 13 above)
Comment (dharmendrak) [reply build.gradle]: reworded file structure

File: .gitignore [hunk 14]
```diff
@@ -0,0 +1,14 @@
+.gradle
```
Comment (timja-kainos) [inline .gitignore]: quick discussion on the naming of this repo, 
where did `service-authorisation-access-client` come from?

The api this talks to is called `service-auth-provider-api` it is also known as `service-2-service-auth` and `s2s`

I would suggest:
`service-2-service-java-client`
or
`service-auth-provider-java-client`

Thoughts anyone?

File: build.gradle (hunk 11 above)
Comment (dharmendrak) [reply build.gradle]: upgraded to latest version 'Edgware.RELEASE'

File: .gitignore (hunk 14 above)
Comment (rkondratowicz) [reply .gitignore]: I personally like the second one more

Comment (timja-kainos) [reply .gitignore]: I'm happy with that, @dharmendrak ?

File: build.gradle [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+            artifact sourcesJar
+            artifact javadocJar
+            groupId project.group
+            artifactId 'service-authorisation-access-client'
+            version project.version
+
+            pom.withXml {
+                def root = asNode()
+                root.appendNode('description', 'Client library for communicating with idam service authorisation access')
+                root.appendNode('name', 'Service authorisation access client')
```
Comment (timja-kainos) [inline build.gradle]: needs an update on this line, the one above and the one below

File: build.gradle [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+bintray {
+    user = System.getenv('BINTRAY_USER')
+    key = System.getenv('BINTRAY_KEY')
+    publications = ['Main']
+    publish = true
+    pkg {
+        repo = 'hmcts-maven'
+        name = 'service-authorisation-access-client'
```
Comment (timja-kainos) [inline build.gradle]: and this

File: build.gradle [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+        userOrg = 'hmcts'
+        licenses = ['MIT']
+        vcsUrl = 'https://github.com/hmcts/service-authorisation-access-client'
```
Comment (timja-kainos) [inline build.gradle]: and here

File: build.gradle [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+    id 'io.spring.dependency-management' version '1.0.3.RELEASE'
+    id 'org.springframework.boot' version '1.5.9.RELEASE' apply true
+    id 'com.jfrog.bintray' version '1.8.0' apply true
+    id 'java-library' apply true
+    id 'maven' apply true
+    id 'maven-publish' apply true
+}
+
+group 'uk.gov.hmcts.reform'
+version '1.0.0'
```
Comment (rkondratowicz) [inline build.gradle]: perhaps '0.0.1'?

File: src/main/resources/META-INF/spring.factories [hunk 19]
```diff
@@ -0,0 +1 @@
+org.springframework.boot.autoconfigure.EnableAutoConfiguration=uk.gov.hmcts.reform.authorisation.ServiceAuthorisationApi
```
Comment (rkondratowicz) [inline src/main/resources/META-INF/spring.factories]: is this file needed?

File: src/main/java/uk/gov/hmcts/reform/authorisation/generators/ServiceAuthTokenGenerator.java [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+            final ServiceAuthorisationApi serviceAuthorisationApi
+    ) {
+        this.secret = secret;
+        this.microService = microService;
+        this.serviceAuthorisationApi = serviceAuthorisationApi;
+        this.googleAuthenticator = new GoogleAuthenticator();
+    }
+
+    @Override
+    public String generate() {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/authorisation/generators/ServiceAuthTokenGenerator.java]: Is it token or auth headers? Add javadoc.

File: src/main/resources/META-INF/spring.factories (hunk 19 above)
Comment (dharmendrak) [reply src/main/resources/META-INF/spring.factories]: it is needed for feign client

File: src/main/java/uk/gov/hmcts/reform/authorisation/generators/ServiceAuthTokenGenerator.java [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+        final String oneTimePassword = format("%06d", googleAuthenticator.getTotpPassword(secret));
+        return "bearer " + serviceAuthorisationApi.serviceToken(this.microService, oneTimePassword);
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/authorisation/generators/ServiceAuthTokenGenerator.java]: Should be capital 'B' I think.
If it's actually needed, IMHO it should be just a token. @timja-kainos ?

File: build.gradle (hunk 15 above)
Comment (dharmendrak) [reply build.gradle]: done

File: src/main/java/uk/gov/hmcts/reform/authorisation/generators/ServiceAuthTokenGenerator.java (hunk 21 above)
Comment (timja-kainos) [reply src/main/java/uk/gov/hmcts/reform/authorisation/generators/ServiceAuthTokenGenerator.java]: header parsing should be done case insensitively, http2 is lower case I believe. 
imo it is correct as is although convention is usually for it to be upper case.

but...
Imo generate shouldn't add the `Bearer <space>`,
Bearer isn't part of the token, it is part of the auth header format

If this was `ServiceAuthHeaderGenerator` then it would be appropriate 😉

File: src/main/java/uk/gov/hmcts/reform/authorisation/ServiceAuthorisationApi.java [hunk 22]
```diff
@@ -0,0 +1,15 @@
+package uk.gov.hmcts.reform.authorisation;
+
+import org.springframework.cloud.netflix.feign.EnableFeignClients;
+import org.springframework.cloud.netflix.feign.FeignClient;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestParam;
+
+@EnableFeignClients
+@FeignClient(name = "idam-s2s-auth", url = "${idam.s2s-auth.url}")
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/authorisation/ServiceAuthorisationApi.java]: Allow users of this lib to instantiate the client with passed url. If I understand correctly, this is taked from env vars?

Also, add sample usage to README.

File: gradle/wrapper/gradle-wrapper.properties [hunk 23]
```diff
@@ -0,0 +1,6 @@
+#Mon Dec 04 11:50:49 GMT 2017
+distributionBase=GRADLE_USER_HOME
+distributionPath=wrapper/dists
+zipStoreBase=GRADLE_USER_HOME
+zipStorePath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.3.1-all.zip
```
Comment (timja-kainos) [inline gradle/wrapper/gradle-wrapper.properties]: gradle 4.4 is out fyi, can upgrade separately though

File: src/main/java/uk/gov/hmcts/reform/authorisation/ServiceAuthorisationApi.java (hunk 22 above)
Comment (timja-kainos) [reply src/main/java/uk/gov/hmcts/reform/authorisation/ServiceAuthorisationApi.java]: how?
this is where the feign magic comes in, set a url in your properties / env vars and it will be resolved.

Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/authorisation/ServiceAuthorisationApi.java]: something like
```
ServiceAuthorisationApi api = 
  Feign
    .builder()                 
    .target(ServiceAuthorisationApi.class, "https://whatever.com");

```

File: src/main/java/uk/gov/hmcts/reform/authorisation/generators/ServiceAuthTokenGenerator.java (hunk 21 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/authorisation/generators/ServiceAuthTokenGenerator.java]: Cool, let's return just the token as the class name suggest and add `"Bearer "` prefix when constructing a header to s2s api.

File: src/main/java/uk/gov/hmcts/reform/authorisation/ServiceAuthorisationApi.java (hunk 22 above)
Comment (timja-kainos) [reply src/main/java/uk/gov/hmcts/reform/authorisation/ServiceAuthorisationApi.java]: that should probably work already?
but when we have someone with that use case it can be looked at.
Currently all apps I am aware of use spring boot, so not sure why we would try and do that.

**Comment by rkondratowicz** [comment]
In the future, please create a pull request with boilerplate changes (like gitignore, gradle, checkstyle, pmd, etc) separately.

File: src/main/java/uk/gov/hmcts/reform/authorisation/ServiceAuthorisationApi.java (hunk 22 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/authorisation/ServiceAuthorisationApi.java]: Maybe, but all the Feign stuff should be internal to this library imho.

I think it would be much clearer if spring boot apps using this library would define a bean which in their configuration where this class is created with explicitly passed url (albeit retrieved from env vars :))

File: README.md [hunk 24]
```diff
@@ -1 +1,78 @@
 # service-auth-provider-java-client
+
+This is the backend service for Civil Money Claims.
```
Comment (rkondratowicz) [inline README.md]: copy pasted? :)

File: README.md [hunk 25]
```diff
@@ -1 +1,78 @@
...
+   
+       @Bean
+       public AuthTokenGenerator cachedServiceAuthTokenGenerator(
+               @Qualifier("serviceAuthTokenGenerator") final AuthTokenGenerator serviceAuthTokenGenerator,
+               @Value("${idam.s2s-auth.tokenTimeToLiveInSeconds:14400}") final int ttl) {
+           return new CachedServiceAuthTokenGenerator(serviceAuthTokenGenerator, ttl);
+       }
+   }
+    
+```
```
Comment (rkondratowicz) [inline README.md]: just explain how to create the client and how to retrieve a token.

File: src/main/java/uk/gov/hmcts/reform/authorisation/ServiceAuthorisationApi.java (hunk 22 above)
Comment (timja-kainos) [reply src/main/java/uk/gov/hmcts/reform/authorisation/ServiceAuthorisationApi.java]: not too worried either way here.

File: src/main/java/uk/gov/hmcts/reform/authorisation/generators/ServiceAuthTokenGenerator.java [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+        this.microService = microService;
+        this.serviceAuthorisationApi = serviceAuthorisationApi;
+        this.googleAuthenticator = new GoogleAuthenticator();
+    }
+
+    /***
+     *
+     * @return Service Auth token for the micro service
+     */
+
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/authorisation/generators/ServiceAuthTokenGenerator.java]: remove empty line

File: src/main/java/uk/gov/hmcts/reform/authorisation/generators/ServiceAuthTokenGenerator.java [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+    @Override
+    public String generate() {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/authorisation/generators/ServiceAuthTokenGenerator.java]: Add a test

**Comment by rkondratowicz** [comment]
the readme comment about caching is not true

File: .travis.yml [hunk 28]
```diff
@@ -0,0 +1,4 @@
+language: java
+jdk:
+- oraclejdk7
+- openjdk6
```
Comment (rkondratowicz) [inline .travis.yml]: is this needed?

Comment (dharmendrak) [reply .travis.yml]: removed

Comment (dharmendrak) [reply .travis.yml]: @timja-kainos please put api key

Comment (timja-kainos) [reply .travis.yml]: could this be a separate PR please?

File: src/test/java/uk/gov/hmcts/reform/authorisation/generators/CachedServiceAuthTokenGeneratorTest.java [hunk 29]
```diff
@@ -0,0 +1,32 @@
...
+
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+public class CachedServiceAuthTokenGeneratorTest {
+
+    private final AuthTokenGenerator delegate = Mockito.mock(AuthTokenGenerator.class);
+
+    @Test
+    public void shouldReuserOldTokenIfTimeHasNotExpired() {
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/authorisation/generators/CachedServiceAuthTokenGeneratorTest.java]: typo in 'reuser'

Comment (dharmendrak) [reply src/test/java/uk/gov/hmcts/reform/authorisation/generators/CachedServiceAuthTokenGeneratorTest.java]: done
